File: compilers/python/runar_compiler/frontend/validator.py
"""Validation pass for the Runar compiler.

Checks the AST against language subset constraints WITHOUT modifying it.
Direct port of ``compilers/go/frontend/validator.go``.
"""

from __future__ import annotations

from dataclasses import dataclass, field

from runar_compiler.frontend.ast_nodes import (
    AssignmentStmt,
    BigIntLiteral,
    BinaryExpr,
    BoolLiteral,
    CallExpr,
    ContractNode,
    CustomType,
    DecrementExpr,
    Expression,
    ExpressionStmt,
    FixedArrayType,
    ForStmt,
    Identifier,
    IfStmt,
    IncrementExpr,
    IndexAccessExpr,
    MemberExpr,
    PrimitiveType,
    PropertyAccessExpr,
    ReturnStmt,
    SourceLocation,
    Statement,
    TernaryExpr,
    TypeNode,
    UnaryExpr,
    VariableDeclStmt,
)


# ---------------------------------------------------------------------------
# Public API
# ---------------------------------------------------------------------------

@dataclass
class ValidationResult:
    """Output of the validation pass."""

    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)


def validate(contract: ContractNode) -> ValidationResult:
    """Validate a Runar AST against language subset constraints.

    Does NOT modify the AST; only reports errors and warnings.
    """
    ctx = _ValidationContext(contract=contract)

    ctx.validate_properties()
    ctx.validate_constructor()
    ctx.validate_methods()
    ctx.check_no_recursion()

    return ValidationResult(errors=ctx.errors, warnings=ctx.warnings)


# ---------------------------------------------------------------------------
# Valid property types
# ---------------------------------------------------------------------------

_VALID_PROP_TYPES: frozenset[str] = frozenset({
    "bigint",
    "boolean",
    "ByteString",
    "PubKey",
    "Sig",
    "Sha256",
    "Ripemd160",
    "Addr",
    "SigHashPreimage",
    "RabinSig",
    "RabinPubKey",
    "Point",
})


# ---------------------------------------------------------------------------
# Validation context
# ---------------------------------------------------------------------------

@dataclass
class _ValidationContext:
    contract: ContractNode
    errors: list[str] = field(default_factory=list)
    warnings: list[str] = field(default_factory=list)

    def _add_error(self, msg: str) -> None:
        self.errors.append(msg)

    # -------------------------------------------------------------------
    # Property validation
    # -------------------------------------------------------------------

    def validate_properties(self) -> None:
        for prop in self.contract.properties:
            self._validate_property_type(prop.type, prop.source_location)

            # V27: txPreimage is an implicit property of StatefulSmartContract
            if self.contract.parent_class == "StatefulSmartContract" and prop.name == "txPreimage":
                self._add_error(
                    "'txPreimage' is an implicit property of StatefulSmartContract "
                    "and must not be declared"
                )

        # SmartContract requires all properties to be readonly
        if self.contract.parent_class == "SmartContract":
            for prop in self.contract.properties:
                if not prop.readonly:
                    self._add_error(
                        f"Property '{prop.name}' in SmartContract must be declared readonly"
                    )

        # V26: Warn if StatefulSmartContract has no mutable properties
        if self.contract.parent_class == "StatefulSmartContract":
            has_mutable = any(not p.readonly for p in self.contract.properties)
            if not has_mutable:
                self.warnings.append(
                    "StatefulSmartContract has no mutable properties; "
                    "consider using SmartContract instead"
                )

    def _validate_property_type(self, t: TypeNode | None, loc: SourceLocation) -> None:
        if t is None:
            return
        if isinstance(t, PrimitiveType):
            if t.name not in _VALID_PROP_TYPES:
                if t.name == "void":
                    self._add_error(
                        f"property type 'void' is not valid at {loc.file}:{loc.line}"
                    )
        elif isinstance(t, FixedArrayType):
            if t.length <= 0:
                self._add_error(
                    f"FixedArray length must be a positive integer at {loc.file}:{loc.line}"
                )
            self._validate_property_type(t.element, loc)
        elif isinstance(t, CustomType):
            self._add_error(
                f"unsupported type '{t.name}' in property declaration at {loc.file}:{loc.line}"
            )

    # -------------------------------------------------------------------
    # Constructor validation
    # -------------------------------------------------------------------

    def validate_constructor(self) -> None:
        ctor = self.contract.constructor
        prop_names: set[str] = {p.name for p in self.contract.properties}

        # Check super() as first statement
        if len(ctor.body) == 0:
            self._add_error("constructor must call super() as its first statement")
            return

        if not _is_super_call(ctor.body[0]):
            self._add_error("constructor must call super() as its first statement")

        # Check all properties are assigned
        assigned_props: set[str] = set()
        for stmt in ctor.body:
            if isinstance(stmt, AssignmentStmt):
                if isinstance(stmt.target, PropertyAccessExpr):
                    assigned_props.add(stmt.target.property)

        # Properties with initializers don't need constructor assignments
        props_with_init = {
            p.name for p in self.contract.properties if p.initializer is not None
        }

        for name in prop_names:
            if name not in assigned_props and name not in props_with_init:
                self._add_error(
                    f"property '{name}' must be assigned in the constructor"
                )

        # Validate constructor body
        for stmt in ctor.body:
            self._validate_statement(stmt)

    # -------------------------------------------------------------------
    # Method validation
    # -------------------------------------------------------------------

    def validate_methods(self) -> None:
        for method in self.contract.methods:
            self._validate_method(method)

    def _validate_method(self, method) -> None:
        # Public methods must end with assert() (unless StatefulSmartContract,
        # where the compiler auto-injects the final assert)
        if (
            method.visibility == "public"
            and self.contract.parent_class != "StatefulSmartContract"
        ):
            if not _ends_with_assert(method.body):
                self._add_error(
                    f"public method '{method.name}' must end with an assert() call"
                )

        # V24/V25: Warn on manual preimage/state-script boilerplate in StatefulSmartContract
        if self.contract.parent_class == "StatefulSmartContract" and method.visibility == "public":
            _warn_manual_preimage_usage(method, self.warnings)

        # Validate statements
        for stmt in method.body:
            self._validate_statement(stmt)

    # -------------------------------------------------------------------
    # Statement validation
    # -------------------------------------------------------------------

    def _validate_statement(self, stmt: Statement) -> None:
        if isinstance(stmt, VariableDeclStmt):
            self._validate_expression(stmt.init)
        elif isinstance(stmt, AssignmentStmt):
            self._validate_expression(stmt.target)
            self._validate_expression(stmt.value)
        elif isinstance(stmt, IfStmt):
            self._validate_expression(stmt.condition)
            for st in stmt.then:
                self._validate_statement(st)
            for st in stmt.else_:
                self._validate_statement(st)
        elif isinstance(stmt, ForStmt):
            self._validate_for_statement(stmt)
        elif isinstance(stmt, ExpressionStmt):
            self._validate_expression(stmt.expr)
        elif isinstance(stmt, ReturnStmt):
            if stmt.value is not None:
                self._validate_expression(stmt.value)

    def _validate_for_statement(self, stmt: ForStmt) -> None:
        self._validate_expression(stmt.condition)

        # Check constant bounds
        if isinstance(stmt.condition, BinaryExpr):
            if not _is_compile_time_constant(stmt.condition.right):
                self._add_error("for loop bound must be a compile-time constant")

        self._validate_expression(stmt.init.init)
        for s in stmt.body:
            self._validate_statement(s)

    # -------------------------------------------------------------------
    # Expression validation
    # -------------------------------------------------------------------

    def _validate_expression(self, expr: Expression | None) -> None:
        if expr is None:
            return
        if isinstance(expr, BinaryExpr):
            self._validate_expression(expr.left)
            self._validate_expression(expr.right)
        elif isinstance(expr, UnaryExpr):
            self._validate_expression(expr.operand)
        elif isinstance(expr, CallExpr):
            self._validate_expression(expr.callee)
            for arg in expr.args:
                self._validate_expression(arg)
        elif isinstance(expr, MemberExpr):
            self._validate_expression(expr.object)
        elif isinstance(expr, TernaryExpr):
            self._validate_expression(expr.condition)
            self._validate_expression(expr.consequent)
            self._validate_expression(expr.alternate)
        elif isinstance(expr, IndexAccessExpr):
            self._validate_expression(expr.object)
            self._validate_expression(expr.index)
        elif isinstance(expr, IncrementExpr):
            self._validate_expression(expr.operand)
        elif isinstance(expr, DecrementExpr):
            self._validate_expression(expr.operand)

    # -------------------------------------------------------------------
    # Recursion detection
    # -------------------------------------------------------------------

    def check_no_recursion(self) -> None:
        call_graph: dict[str, set[str]] = {}
        method_names: set[str] = set()

        for method in self.contract.methods:
            method_names.add(method.name)
            calls: set[str] = set()
            _collect_method_calls(method.body, calls)
            call_graph[method.name] = calls

        # Check for cycles using DFS
        for method in self.contract.methods:
            visited: set[str] = set()
            stack: set[str] = set()
            if _has_cycle(method.name, call_graph, method_names, visited, stack):
                self._add_error(
                    f"recursion detected: method '{method.name}' calls itself "
                    f"directly or indirectly"
                )


# ---------------------------------------------------------------------------
# Helper functions
# ---------------------------------------------------------------------------

def _is_super_call(stmt: Statement) -> bool:
    if not isinstance(stmt, ExpressionStmt):
        return False
    if not isinstance(stmt.expr, CallExpr):
        return False
    callee = stmt.expr.callee
    # Accept both Identifier("super") and MemberExpr(Identifier("super"), "")
    if isinstance(callee, Identifier):
        return callee.name == "super"
    if isinstance(callee, MemberExpr):
        return isinstance(callee.object, Identifier) and callee.object.name == "super"
    return False


def _ends_with_assert(body: list[Statement]) -> bool:
    if len(body) == 0:
        return False
    last = body[-1]

    if isinstance(last, ExpressionStmt):
        return _is_assert_call(last.expr)

    if isinstance(last, IfStmt):
        then_ends = _ends_with_assert(last.then)
        else_ends = len(last.else_) > 0 and _ends_with_assert(last.else_)
        return then_ends and else_ends

    return False


def _is_assert_call(expr: Expression | None) -> bool:
    if not isinstance(expr, CallExpr):
        return False
    if not isinstance(expr.callee, Identifier):
        return False
    return expr.callee.name == "assert"


def _is_compile_time_constant(expr: Expression | None) -> bool:
    if expr is None:
        return False
    if isinstance(expr, BigIntLiteral):
        return True
    if isinstance(expr, BoolLiteral):
        return True
    if isinstance(expr, Identifier):
        return True  # trust it's a const
    if isinstance(expr, UnaryExpr):
        if expr.op == "-":
            return _is_compile_time_constant(expr.operand)
    return False


# ---------------------------------------------------------------------------
# V24/V25: warn on manual preimage/state-script usage
# ---------------------------------------------------------------------------

def _warn_manual_preimage_usage(method, warnings: list[str]) -> None:
    """Walk method body and warn on checkPreimage() or this.getStateScript() calls."""

    def visitor(expr: Expression) -> None:
        if isinstance(expr, CallExpr):
            # V24: bare checkPreimage(...) call
            if isinstance(expr.callee, Identifier) and expr.callee.name == "checkPreimage":
                warnings.append(
                    f"StatefulSmartContract auto-injects checkPreimage(); calling it manually "
                    f"in '{method.name}' will cause a duplicate verification"
                )
            # V24: this.checkPreimage(...) call via PropertyAccessExpr or MemberExpr
            callee_prop = _callee_property(expr.callee)
            if callee_prop == "checkPreimage":
                warnings.append(
                    f"StatefulSmartContract auto-injects checkPreimage(); calling it manually "
                    f"in '{method.name}' will cause a duplicate verification"
                )
            # V25: this.getStateScript() call
            if callee_prop == "getStateScript":
                warnings.append(
                    f"StatefulSmartContract auto-injects state continuation; calling "
                    f"getStateScript() manually in '{method.name}' is redundant"
                )

    _walk_expressions_in_body(method.body, visitor)


def _callee_property(callee: Expression | None) -> str | None:
    """Return the property name if callee is a property access (PropertyAccessExpr or MemberExpr)."""
    if callee is None:
        return None
    if isinstance(callee, PropertyAccessExpr):
        return callee.property
    if isinstance(callee, MemberExpr):
        return callee.property
    return None


def _walk_expressions_in_body(stmts: list[Statement], visitor) -> None:
    for stmt in stmts:
        _walk_expressions_in_stmt(stmt, visitor)


def _walk_expressions_in_stmt(stmt: Statement, visitor) -> None:
    if isinstance(stmt, ExpressionStmt):
        _walk_expr(stmt.expr, visitor)
    elif isinstance(stmt, VariableDeclStmt):
        _walk_expr(stmt.init, visitor)
    elif isinstance(stmt, AssignmentStmt):
        _walk_expr(stmt.target, visitor)
        _walk_expr(stmt.value, visitor)
    elif isinstance(stmt, IfStmt):
        _walk_expr(stmt.condition, visitor)
        _walk_expressions_in_body(stmt.then, visitor)
        _walk_expressions_in_body(stmt.else_, visitor)
    elif isinstance(stmt, ForStmt):
        _walk_expr(stmt.condition, visitor)
        _walk_expressions_in_body(stmt.body, visitor)
    elif isinstance(stmt, ReturnStmt):
        if stmt.value is not None:
            _walk_expr(stmt.value, visitor)


def _walk_expr(expr: Expression | None, visitor) -> None:
    if expr is None:
        return
    visitor(expr)
    if isinstance(expr, BinaryExpr):
        _walk_expr(expr.left, visitor)
        _walk_expr(expr.right, visitor)
    elif isinstance(expr, UnaryExpr):
        _walk_expr(expr.operand, visitor)
    elif isinstance(expr, CallExpr):
        _walk_expr(expr.callee, visitor)
        for arg in expr.args:
            _walk_expr(arg, visitor)
    elif isinstance(expr, MemberExpr):
        _walk_expr(expr.object, visitor)
    elif isinstance(expr, TernaryExpr):
        _walk_expr(expr.condition, visitor)
        _walk_expr(expr.consequent, visitor)
        _walk_expr(expr.alternate, visitor)
    elif isinstance(expr, IndexAccessExpr):
        _walk_expr(expr.object, visitor)
        _walk_expr(expr.index, visitor)
    elif isinstance(expr, IncrementExpr):
        _walk_expr(expr.operand, visitor)
    elif isinstance(expr, DecrementExpr):
        _walk_expr(expr.operand, visitor)


# ---------------------------------------------------------------------------
# Recursion detection helpers
# ---------------------------------------------------------------------------

def _collect_method_calls(stmts: list[Statement], calls: set[str]) -> None:
    for stmt in stmts:
        _collect_method_calls_in_stmt(stmt, calls)


def _collect_method_calls_in_stmt(stmt: Statement, calls: set[str]) -> None:
    if isinstance(stmt, ExpressionStmt):
        _collect_method_calls_in_expr(stmt.expr, calls)
    elif isinstance(stmt, VariableDeclStmt):
        _collect_method_calls_in_expr(stmt.init, calls)
    elif isinstance(stmt, AssignmentStmt):
        _collect_method_calls_in_expr(stmt.target, calls)
        _collect_method_calls_in_expr(stmt.value, calls)
    elif isinstance(stmt, IfStmt):
        _collect_method_calls_in_expr(stmt.condition, calls)
        _collect_method_calls(stmt.then, calls)
        _collect_method_calls(stmt.else_, calls)
    elif isinstance(stmt, ForStmt):
        _collect_method_calls_in_expr(stmt.condition, calls)
        _collect_method_calls(stmt.body, calls)
    elif isinstance(stmt, ReturnStmt):
        if stmt.value is not None:
            _collect_method_calls_in_expr(stmt.value, calls)


def _collect_method_calls_in_expr(expr: Expression | None, calls: set[str]) -> None:
    if expr is None:
        return
    if isinstance(expr, CallExpr):
        if isinstance(expr.callee, PropertyAccessExpr):
            calls.add(expr.callee.property)
        if isinstance(expr.callee, MemberExpr):
            if isinstance(expr.callee.object, Identifier) and expr.callee.object.name == "this":
                calls.add(expr.callee.property)
        _collect_method_calls_in_expr(expr.callee, calls)
        for arg in expr.args:
            _collect_method_calls_in_expr(arg, calls)
    elif isinstance(expr, BinaryExpr):
        _collect_method_calls_in_expr(expr.left, calls)
        _collect_method_calls_in_expr(expr.right, calls)
    elif isinstance(expr, UnaryExpr):
        _collect_method_calls_in_expr(expr.operand, calls)
    elif isinstance(expr, MemberExpr):
        _collect_method_calls_in_expr(expr.object, calls)
    elif isinstance(expr, TernaryExpr):
        _collect_method_calls_in_expr(expr.condition, calls)
        _collect_method_calls_in_expr(expr.consequent, calls)
        _collect_method_calls_in_expr(expr.alternate, calls)
    elif isinstance(expr, IndexAccessExpr):
        _collect_method_calls_in_expr(expr.object, calls)
        _collect_method_calls_in_expr(expr.index, calls)
    elif isinstance(expr, IncrementExpr):
        _collect_method_calls_in_expr(expr.operand, calls)
    elif isinstance(expr, DecrementExpr):
        _collect_method_calls_in_expr(expr.operand, calls)


def _has_cycle(
    name: str,
    call_graph: dict[str, set[str]],
    method_names: set[str],
    visited: set[str],
    stack: set[str],
) -> bool:
    if name in stack:
        return True
    if name in visited:
        return False
    visited.add(name)
    stack.add(name)

    for callee in call_graph.get(name, set()):
        if callee in method_names:
            if _has_cycle(callee, call_graph, method_names, visited, stack):
                return True

    stack.discard(name)
    return False

File: compilers/ruby/lib/runar_compiler/frontend/validator.rb
# frozen_string_literal: true

# Validation pass for the Runar compiler.
#
# Checks the AST against language subset constraints WITHOUT modifying it.
# Direct port of compilers/python/runar_compiler/frontend/validator.py.

require_relative "ast_nodes"
require_relative "diagnostic"

module RunarCompiler
  module Frontend
    # Output of the validation pass.
    class ValidationResult
      attr_reader :errors, :warnings

      def initialize(errors: [], warnings: [])
        @errors = errors
        @warnings = warnings
      end

      # Return formatted error messages as plain strings.
      def error_strings
        @errors.map(&:format_message)
      end

      # Return formatted warning messages as plain strings.
      def warning_strings
        @warnings.map(&:format_message)
      end
    end

    # Valid property types (excluding void).
    VALID_PROP_TYPES = %w[
      bigint
      boolean
      ByteString
      PubKey
      Sig
      Sha256
      Ripemd160
      Addr
      SigHashPreimage
      RabinSig
      RabinPubKey
      Point
    ].to_set.freeze

    # Validate a Runar AST against language subset constraints.
    #
    # Does NOT modify the AST; only reports errors and warnings.
    #
    # @param contract [ContractNode]
    # @return [ValidationResult]
    def self.validate(contract)
      ctx = ValidationContext.new(contract)

      ctx.validate_properties
      ctx.validate_constructor
      ctx.validate_methods
      ctx.check_no_recursion

      ValidationResult.new(errors: ctx.errors, warnings: ctx.warnings)
    end

    # @api private
    class ValidationContext
      attr_reader :errors, :warnings

      def initialize(contract)
        @contract = contract
        @errors = []
        @warnings = []
      end

      # -------------------------------------------------------------------
      # Property validation
      # -------------------------------------------------------------------

      def validate_properties
        @contract.properties.each do |prop|
          validate_property_type(prop.type, prop.source_location)

          # V27: txPreimage is an implicit property of StatefulSmartContract
          if @contract.parent_class == "StatefulSmartContract" && prop.name == "txPreimage"
            add_error(
              "'txPreimage' is an implicit property of StatefulSmartContract " \
              "and must not be declared",
              loc: prop.source_location
            )
          end
        end

        # SmartContract requires all properties to be readonly
        if @contract.parent_class == "SmartContract"
          @contract.properties.each do |prop|
            unless prop.readonly
              add_error(
                "Property '#{prop.name}' in SmartContract must be declared readonly",
                loc: prop.source_location
              )
            end
          end
        end

        # V26: Warn if StatefulSmartContract has no mutable properties
        if @contract.parent_class == "StatefulSmartContract"
          has_mutable = @contract.properties.any? { |p| !p.readonly }
          unless has_mutable
            @warnings << Diagnostic.new(
              message: "StatefulSmartContract has no mutable properties; " \
                       "consider using SmartContract instead",
              severity: Severity::WARNING,
              loc: @contract.constructor.source_location
            )
          end
        end
      end

      # -------------------------------------------------------------------
      # Constructor validation
      # -------------------------------------------------------------------

      def validate_constructor
        ctor = @contract.constructor
        prop_names = @contract.properties.map(&:name).to_set

        # Check super() as first statement
        if ctor.body.empty?
          add_error("constructor must call super() as its first statement", loc: ctor.source_location)
          return
        end

        unless super_call?(ctor.body[0])
          add_error("constructor must call super() as its first statement", loc: ctor.source_location)
        end

        # Check all properties are assigned
        assigned_props = Set.new
        ctor.body.each do |stmt|
          if stmt.is_a?(AssignmentStmt) && stmt.target.is_a?(PropertyAccessExpr)
            assigned_props.add(stmt.target.property)
          end
        end

        # Properties with initializers don't need constructor assignments
        props_with_init = @contract.properties
          .select { |p| !p.initializer.nil? }
          .map(&:name)
          .to_set

        prop_names.each do |name|
          if !assigned_props.include?(name) && !props_with_init.include?(name)
            add_error(
              "property '#{name}' must be assigned in the constructor",
              loc: ctor.source_location
            )
          end
        end

        # Validate constructor body
        ctor.body.each { |stmt| validate_statement(stmt) }
      end

      # -------------------------------------------------------------------
      # Method validation
      # -------------------------------------------------------------------

      def validate_methods
        @contract.methods.each { |method| validate_method(method) }
      end

      # -------------------------------------------------------------------
      # Recursion detection
      # -------------------------------------------------------------------

      def check_no_recursion
        call_graph = {}
        method_names = Set.new

        @contract.methods.each do |method|
          method_names.add(method.name)
          calls = Set.new
          collect_method_calls(method.body, calls)
          call_graph[method.name] = calls
        end

        # Check for cycles using DFS
        @contract.methods.each do |method|
          visited = Set.new
          stack = Set.new
          if has_cycle?(method.name, call_graph, method_names, visited, stack)
            add_error(
              "recursion detected: method '#{method.name}' calls itself " \
              "directly or indirectly",
              loc: method.source_location
            )
          end
        end
      end

      private

      def add_error(msg, loc: nil)
        @errors << Diagnostic.new(message: msg, severity: Severity::ERROR, loc: loc)
      end

      # -------------------------------------------------------------------
      # Property type validation (private)
      # -------------------------------------------------------------------

      def validate_property_type(type_node, loc)
        return if type_node.nil?

        if type_node.is_a?(PrimitiveType)
          unless VALID_PROP_TYPES.include?(type_node.name)
            if type_node.name == "void"
              add_error(
                "property type 'void' is not valid at #{loc.file}:#{loc.line}",
                loc: loc
              )
            end
          end
        elsif type_node.is_a?(FixedArrayType)
          if type_node.length <= 0
            add_error(
              "FixedArray length must be a positive integer at #{loc.file}:#{loc.line}",
              loc: loc
            )
          end
          validate_property_type(type_node.element, loc)
        elsif type_node.is_a?(CustomType)
          add_error(
            "unsupported type '#{type_node.name}' in property declaration at #{loc.file}:#{loc.line}",
            loc: loc
          )
        end
      end

      # -------------------------------------------------------------------
      # Method validation (private)
      # -------------------------------------------------------------------

      def validate_method(method)
        # Public methods must end with assert() (unless StatefulSmartContract,
        # where the compiler auto-injects the final assert)
        if method.visibility == "public" && @contract.parent_class != "StatefulSmartContract"
          unless ends_with_assert?(method.body)
            add_error(
              "public method '#{method.name}' must end with an assert() call",
              loc: method.source_location
            )
          end
        end

        # V24/V25: Warn on manual preimage/state-script boilerplate in StatefulSmartContract
        if @contract.parent_class == "StatefulSmartContract" && method.visibility == "public"
          warn_manual_preimage_usage(method)
        end

        # Validate statements
        method.body.each { |stmt| validate_statement(stmt) }
      end

      # -------------------------------------------------------------------
      # Statement validation (private)
      # -------------------------------------------------------------------

      def validate_statement(stmt)
        case stmt
        when VariableDeclStmt
          validate_expression(stmt.init)
        when AssignmentStmt
          validate_expression(stmt.target)
          validate_expression(stmt.value)
        when IfStmt
          validate_expression(stmt.condition)
          stmt.then.each { |s| validate_statement(s) }
          stmt.else_.each { |s| validate_statement(s) }
        when ForStmt
          validate_for_statement(stmt)
        when ExpressionStmt
          validate_expression(stmt.expr)
        when ReturnStmt
          validate_expression(stmt.value) unless stmt.value.nil?
        end
      end

      def validate_for_statement(stmt)
        validate_expression(stmt.condition)

        # Check constant bounds
        if stmt.condition.is_a?(BinaryExpr)
          unless compile_time_constant?(stmt.condition.right)
            add_error("for loop bound must be a compile-time constant")
          end
        end

        validate_expression(stmt.init.init)
        stmt.body.each { |s| validate_statement(s) }
      end

      # -------------------------------------------------------------------
      # Expression validation (private)
      # -------------------------------------------------------------------

      def validate_expression(expr)
        return if expr.nil?

        case expr
        when BinaryExpr
          validate_expression(expr.left)
          validate_expression(expr.right)
        when UnaryExpr
          validate_expression(expr.operand)
        when CallExpr
          validate_expression(expr.callee)
          # assert() message (2nd arg) is a human-readable string, not hex -- skip validation
          is_assert = expr.callee.is_a?(Identifier) && expr.callee.name == "assert"
          expr.args.each_with_index do |arg, i|
            next if is_assert && i >= 1
            validate_expression(arg)
          end
        when MemberExpr
          validate_expression(expr.object)
        when TernaryExpr
          validate_expression(expr.condition)
          validate_expression(expr.consequent)
          validate_expression(expr.alternate)
        when IndexAccessExpr
          validate_expression(expr.object)
          validate_expression(expr.index)
        when IncrementExpr
          validate_expression(expr.operand)
        when DecrementExpr
          validate_expression(expr.operand)
        when ByteStringLiteral
          val = expr.value
          if val.length > 0
            if val.length.odd?
              add_error(
                "ByteString literal '#{val}' has odd length (#{val.length}) " \
                "\u2014 hex strings must have an even number of characters"
              )
            elsif !val.match?(/\A[0-9a-fA-F]*\z/)
              add_error(
                "ByteString literal '#{val}' contains non-hex characters " \
                "\u2014 only 0-9, a-f, A-F are allowed"
              )
            end
          end
        end
      end

      # -------------------------------------------------------------------
      # Helper: super() call detection
      # -------------------------------------------------------------------

      def super_call?(stmt)
        return false unless stmt.is_a?(ExpressionStmt)
        return false unless stmt.expr.is_a?(CallExpr)

        callee = stmt.expr.callee
        # Accept both Identifier("super") and MemberExpr(Identifier("super"), "")
        if callee.is_a?(Identifier)
          return callee.name == "super"
        end
        if callee.is_a?(MemberExpr)
          return callee.object.is_a?(Identifier) && callee.object.name == "super"
        end

        false
      end

      # -------------------------------------------------------------------
      # Helper: ends_with_assert?
      # -------------------------------------------------------------------

      def ends_with_assert?(body)
        return false if body.empty?

        last = body.last

        if last.is_a?(ExpressionStmt)
          return assert_call?(last.expr)
        end

        if last.is_a?(IfStmt)
          then_ends = ends_with_assert?(last.then)
          else_ends = !last.else_.empty? && ends_with_assert?(last.else_)
          return then_ends && else_ends
        end

        false
      end

      def assert_call?(expr)
        return false unless expr.is_a?(CallExpr)
        return false unless expr.callee.is_a?(Identifier)

        expr.callee.name == "assert"
      end

      # -------------------------------------------------------------------
      # Helper: compile-time constant check
      # -------------------------------------------------------------------

      def compile_time_constant?(expr)
        return false if expr.nil?
        return true if expr.is_a?(BigIntLiteral)
        return true if expr.is_a?(BoolLiteral)
        return true if expr.is_a?(Identifier) # trust it's a const

        if expr.is_a?(UnaryExpr) && expr.op == "-"
          return compile_time_constant?(expr.operand)
        end

        false
      end

      # -------------------------------------------------------------------
      # V24/V25: warn on manual preimage/state-script usage
      # -------------------------------------------------------------------

      def warn_manual_preimage_usage(method)
        method_loc = method.source_location

        visitor = proc do |expr|
          if expr.is_a?(CallExpr)
            # V24: bare checkPreimage(...) call
            if expr.callee.is_a?(Identifier) && expr.callee.name == "checkPreimage"
              @warnings << Diagnostic.new(
                message: "StatefulSmartContract auto-injects checkPreimage(); calling it manually " \
                         "in '#{method.name}' will cause a duplicate verification",
                severity: Severity::WARNING,
                loc: method_loc
              )
            end
            # V24: this.checkPreimage(...) call via PropertyAccessExpr or MemberExpr
            callee_prop = callee_property(expr.callee)
            if callee_prop == "checkPreimage"
              @warnings << Diagnostic.new(
                message: "StatefulSmartContract auto-injects checkPreimage(); calling it manually " \
                         "in '#{method.name}' will cause a duplicate verification",
                severity: Severity::WARNING,
                loc: method_loc
              )
            end
            # V25: this.getStateScript() call
            if callee_prop == "getStateScript"
              @warnings << Diagnostic.new(
                message: "StatefulSmartContract auto-injects state continuation; calling " \
                         "getStateScript() manually in '#{method.name}' is redundant",
                severity: Severity::WARNING,
                loc: method_loc
              )
            end
          end
        end

        walk_expressions_in_body(method.body, visitor)
      end

      def callee_property(callee)
        return nil if callee.nil?
        return callee.property if callee.is_a?(PropertyAccessExpr)
        return callee.property if callee.is_a?(MemberExpr)

        nil
      end

      # -------------------------------------------------------------------
      # Expression tree walkers
      # -------------------------------------------------------------------

      def walk_expressions_in_body(stmts, visitor)
        stmts.each { |stmt| walk_expressions_in_stmt(stmt, visitor) }
      end

      def walk_expressions_in_stmt(stmt, visitor)
        case stmt
        when ExpressionStmt
          walk_expr(stmt.expr, visitor)
        when VariableDeclStmt
          walk_expr(stmt.init, visitor)
        when AssignmentStmt
          walk_expr(stmt.target, visitor)
          walk_expr(stmt.value, visitor)
        when IfStmt
          walk_expr(stmt.condition, visitor)
          walk_expressions_in_body(stmt.then, visitor)
          walk_expressions_in_body(stmt.else_, visitor)
        when ForStmt
          walk_expr(stmt.condition, visitor)
          walk_expressions_in_body(stmt.body, visitor)
        when ReturnStmt
          walk_expr(stmt.value, visitor) unless stmt.value.nil?
        end
      end

      def walk_expr(expr, visitor)
        return if expr.nil?

        visitor.call(expr)

        case expr
        when BinaryExpr
          walk_expr(expr.left, visitor)
          walk_expr(expr.right, visitor)
        when UnaryExpr
          walk_expr(expr.operand, visitor)
        when CallExpr
          walk_expr(expr.callee, visitor)
          expr.args.each { |arg| walk_expr(arg, visitor) }
        when MemberExpr
          walk_expr(expr.object, visitor)
        when TernaryExpr
          walk_expr(expr.condition, visitor)
          walk_expr(expr.consequent, visitor)
          walk_expr(expr.alternate, visitor)
        when IndexAccessExpr
          walk_expr(expr.object, visitor)
          walk_expr(expr.index, visitor)
        when IncrementExpr
          walk_expr(expr.operand, visitor)
        when DecrementExpr
          walk_expr(expr.operand, visitor)
        end
      end

      # -------------------------------------------------------------------
      # Recursion detection helpers
      # -------------------------------------------------------------------

      def collect_method_calls(stmts, calls)
        stmts.each { |stmt| collect_method_calls_in_stmt(stmt, calls) }
      end

      def collect_method_calls_in_stmt(stmt, calls)
        case stmt
        when ExpressionStmt
          collect_method_calls_in_expr(stmt.expr, calls)
        when VariableDeclStmt
          collect_method_calls_in_expr(stmt.init, calls)
        when AssignmentStmt
          collect_method_calls_in_expr(stmt.target, calls)
          collect_method_calls_in_expr(stmt.value, calls)
        when IfStmt
          collect_method_calls_in_expr(stmt.condition, calls)
          collect_method_calls(stmt.then, calls)
          collect_method_calls(stmt.else_, calls)
        when ForStmt
          collect_method_calls_in_expr(stmt.condition, calls)
          collect_method_calls(stmt.body, calls)
        when ReturnStmt
          collect_method_calls_in_expr(stmt.value, calls) unless stmt.value.nil?
        end
      end

      def collect_method_calls_in_expr(expr, calls)
        return if expr.nil?

        case expr
        when CallExpr
          if expr.callee.is_a?(PropertyAccessExpr)
            calls.add(expr.callee.property)
          end
          if expr.callee.is_a?(MemberExpr)
            if expr.callee.object.is_a?(Identifier) && expr.callee.object.name == "this"
              calls.add(expr.callee.property)
            end
          end
          collect_method_calls_in_expr(expr.callee, calls)
          expr.args.each { |arg| collect_method_calls_in_expr(arg, calls) }
        when BinaryExpr
          collect_method_calls_in_expr(expr.left, calls)
          collect_method_calls_in_expr(expr.right, calls)
        when UnaryExpr
          collect_method_calls_in_expr(expr.operand, calls)
        when MemberExpr
          collect_method_calls_in_expr(expr.object, calls)
        when TernaryExpr
          collect_method_calls_in_expr(expr.condition, calls)
          collect_method_calls_in_expr(expr.consequent, calls)
          collect_method_calls_in_expr(expr.alternate, calls)
        when IndexAccessExpr
          collect_method_calls_in_expr(expr.object, calls)
          collect_method_calls_in_expr(expr.index, calls)
        when IncrementExpr
          collect_method_calls_in_expr(expr.operand, calls)
        when DecrementExpr
          collect_method_calls_in_expr(expr.operand, calls)
        end
      end

      def has_cycle?(name, call_graph, method_names, visited, stack)
        return true if stack.include?(name)
        return false if visited.include?(name)

        visited.add(name)
        stack.add(name)

        (call_graph[name] || Set.new).each do |callee|
          if method_names.include?(callee)
            return true if has_cycle?(callee, call_graph, method_names, visited, stack)
          end
        end

        stack.delete(name)
        false
      end
    end
  end
end
